File: colores_ascii.dart
import 'dart:io';

// Directorios de archivos de origen
final directorioPugOrigen = './src_pug';
final directorioStylusOrigen = './src_stylus';
final directorioCoffeeScriptOrigen = './src_coffeescript';

// Directorios de archivos de destino
final directorioHtmlDestino = './build_html';
final directorioCssDestino = './build_css';
final directorioJsDestino = './build_js';

// Comandos para compilar archivos Pug, Stylus y CoffeeScript
final comandoCompilarPug = 'pug {} -o {}';
final comandoCompilarStylus = 'stylus {} -o {}';
final comandoCompilarCoffeeScript = 'coffee -c {} -o {}';

// Función para ejecutar un comando en la línea de comandos
Future<void> ejecutarComando(String comando) async {
  final proceso = await Process.start(comando, [], runInShell: true);
  final stdoutStream = proceso.stdout.transform(Utf8Decoder());
  final stderrStream = proceso.stderr.transform(Utf8Decoder());

  await for (final line in stdoutStream) {
    print(line);
  }

  await for (final line in stderrStream) {
    print(line);
  }
}

// Función para observar cambios en archivos
Future<void> observarCambios() async {
  try {
    while (true) {
      await for (final entity in Directory(directorioPugOrigen).list(recursive: true)) {
        if (entity is File && entity.path.endsWith('.pug')) {
          final rutaArchivo = entity.path;
          await ejecutarComando(comandoCompilarPug
              .replaceAll('{}', rutaArchivo)
              .replaceAll('{}', directorioHtmlDestino));
        }
      }

      await for (final entity in Directory(directorioStylusOrigen).list(recursive: true)) {
        if (entity is File && entity.path.endsWith('.styl')) {
          final rutaArchivo = entity.path;
          await ejecutarComando(comandoCompilarStylus
              .replaceAll('{}', rutaArchivo)
              .replaceAll('{}', directorioCssDestino));
        }
      }

      await for (final entity in Directory(directorioCoffeeScriptOrigen).list(recursive: true)) {
        if (entity is File && entity.path.endsWith('.coffee')) {
          final rutaArchivo = entity.path;
          await ejecutarComando(comandoCompilarCoffeeScript
              .replaceAll('{}', rutaArchivo)
              .replaceAll('{}', directorioJsDestino));
        }
      }

      print('Presiona Enter para continuar...');
      await stdin.readLineSync();
    }
  } on FileSystemException catch (e) {
    print('Error: $e');
  } on ProcessException catch (e) {
    print('Error al ejecutar el comando: $e');
  }
}

void main() {
  print('Observando archivos Pug, Stylus y CoffeeScript...');
  observarCambios();
}

File: colores_ascii.lua
-- Módulo para ejecutar comandos en la línea de comandos
function ejecutarComando(comando)
    local handle = io.popen(comando)
    local resultado = handle:read("*a")
    handle:close()
    return resultado
end

-- Directorios de archivos de origen
local directorioPugOrigen = './src_pug'
local directorioStylusOrigen = './src_stylus'
local directorioCoffeeScriptOrigen = './src_coffeescript'

-- Directorios de archivos de destino
local directorioHtmlDestino = './build_html'
local directorioCssDestino = './build_css'
local directorioJsDestino = './build_js'

-- Comandos para compilar archivos Pug, Stylus y CoffeeScript
local comandoCompilarPug = 'pug {} -o {}'
local comandoCompilarStylus = 'stylus {} -o {}'
local comandoCompilarCoffeeScript = 'coffee -c {} -o {}'

-- Función para observar cambios en archivos
function observarCambios()
    while true do
        for archivo in io.popen('find ' .. directorioPugOrigen):lines() do
            if string.match(archivo, "%.pug$") then
                local comando = string.format(comandoCompilarPug, archivo, directorioHtmlDestino)
                local resultado = ejecutarComando(comando)
                if resultado ~= "" then
                    print(resultado)
                else
                    print(string.format('Compilado: %s -> %s', archivo, directorioHtmlDestino))
                end
            end
        end

        for archivo in io.popen('find ' .. directorioStylusOrigen):lines() do
            if string.match(archivo, "%.styl$") then
                local comando = string.format(comandoCompilarStylus, archivo, directorioCssDestino)
                local resultado = ejecutarComando(comando)
                if resultado ~= "" then
                    print(resultado)
                else
                    print(string.format('Compilado: %s -> %s', archivo, directorioCssDestino))
                end
            end
        end

        for archivo in io.popen('find ' .. directorioCoffeeScriptOrigen):lines() do
            if string.match(archivo, "%.coffee$") then
                local comando = string.format(comandoCompilarCoffeeScript, archivo, directorioJsDestino)
                local resultado = ejecutarComando(comando)
                if resultado ~= "" then
                    print(resultado)
                else
                    print(string.format('Compilado: %s -> %s', archivo, directorioJsDestino))
                end
            end
        end

        print("Presiona Enter para continuar...")
        io.read()
    end
end

print('Observando archivos Pug, Stylus y CoffeeScript...')
observarCambios()
